### ISSUE TITLE
[Game] Spielgeschwindigkeit ist an Framerate gekoppelt

### ISSUE THREAD

**Issue by fwatermann** [Programmiermethoden/Dungeon#807]
Da unsere Systeme in der Render-Loop ihr Update erhalten ist die Geschwindigkeit des gesamten Spiels davon abhängig. 

Mögliche Ideen:
- Systeme alle x Frames updaten
- Systeme haben eine eigene Tickrate (z.B DrawSystem jeden Frame, Velocity jeden zweiten, etc...)

**Comment by cagix** [comment]
Noch eine andere Idee: Es wird zwar pro Frame neu gerendert, aber neue Positionen/Aktionen werden nur dann berechnet, wenn es auch eine Tastatureingabe gibt. Dann sind wir eher turn-basiert.

**Comment by cagix** [comment]
@AMatutat @fwatermann Noch eine andere Idee: Es wird zwar pro Frame neu gerendert, aber neue Positionen/Aktionen werden nur dann berechnet, wenn es auch eine Tastatureingabe gibt. Dann sind wir eher turn-basiert.

**Comment by AMatutat** [comment]
> @AMatutat @fwatermann Noch eine andere Idee: Es wird zwar pro Frame neu gerendert, aber neue Positionen/Aktionen werden nur dann berechnet, wenn es auch eine Tastatureingabe gibt. Dann sind wir eher turn-basiert.

Du und dein turn-based :P 

Turn-Based fände ich als Toogle-Option auch echt lustig.

**Comment by AKirsch1** [comment]
Da mir dieses Issue gerade aufgefallen ist, will ich auch einmal was dazu schreiben:

Mich wundert nämlich gerade, will man z.B. die Position wirklich nur jeden zweiten Frame updaten? Denn wenn ich das bei jedem System, abgesehen vom DrawSystem, mache, dann sieht jedes zweite Bild wie das vorherige aus und ich halbiere faktisch die Framerate. Mal abgesehen davon, dass damit die Spielgeschwindigkeit immer noch an die Framerate gekoppelt ist.

Was andere Engines machen, ist die Zeit zwischen zwei Frames zu berechnen. In Unity gibt es zum Beispiel `Time.deltaTime`. Für LibGDX wäre es `Gdx.graphics.getDeltaTime()`. Bei einem Update mutipliziere ich meine Velocity dann mit der DeltaTime und werde damit unabhängig von der Framerate.

Für alles andere, das nur unter bestimmten Bedingungen ausgeführt werden soll, wäre vielleicht ein EventSystem geeigneter. Z.B. beim XPSystem: Wenn eine Entität XP bekommt, wird ein Event gefeuert, welches beim Ausführen überprüft, ob ein Level-Up gemacht werden muss.

**Comment by fwatermann** [comment]
@AKirsch1 Stimmt, so kenne ich das auch. Wir hatten genau darüber auch im Meeting am Mittwoch kurz gesprochen, als zusätzliche Möglichkeit.

**Comment by cagix** [comment]
> @AKirsch1 Stimmt, so kenne ich das auch. Wir hatten genau darüber auch im Meeting am Mittwoch kurz gesprochen, als zusätzliche Möglichkeit.

wolltest du nicht deine praxisphase hier machen? 🤣

**Comment by AMatutat** [comment]
See also #244

**Comment by cagix** [comment]
@AMatutat "thesis"?
